File: Lecture-01-Recursion/Code.cpp
#include <iostream>
#include <vector>

using namespace std;

int minimum02(vector<int> arr, int vidx){
    if(vidx == arr.size()-1){
        return arr[vidx];
    }

    int faith = minimum02(arr,vidx+1);
    int minr =  min(faith,arr[vidx]);
    return minr;
}

int minimum01(vector<int> arr, int vidx){
    if(vidx == arr.size()-1){
        return arr[vidx];
    }

    int min = minimum01(arr,vidx+1);
    return (min < arr[vidx]) ? min : arr[vidx];
}

int maximum02(vector<int> arr, int vidx){
    if(vidx == arr.size()-1){
        return arr[vidx];
    }

    int faith = maximum02(arr,vidx+1);
    int maxr =  max(faith,arr[vidx]);
    return maxr;
}

int maximum01(vector<int> arr, int vidx){
    if(vidx == arr.size()-1){
        return arr[vidx];
    }

    int max = maximum01(arr,vidx+1);
    return (max > arr[vidx]) ? max : arr[vidx];
}

void display(vector<int> arr){
    for(int i=0;i<arr.size();i++){
        cout<< arr[i]<<" ";
    }
}

//set2 --------------------------------
int powerBtr(int n, int p){
    if(n == 0){
        return 0;
    }
    
    if(p == 0){
        return 1;
    }

    int faith = powerBtr(n, p/2);
    return (p%2 == 0)? faith * faith : faith * faith * n;
}

int power(int n , int p){
    if(n == 0){
        return 0;
    }
    if(p == 0){
        return 1;
    }

    int faith = power(n , p-1);
    return n * faith;
}

int fact(int n){
    if(n == 0){
        return 1;
    }

    int smalAns = fact(n-1);
    return n * smalAns;
}
//set1 --------------------------------------------

void set2(){
    int n;
    cin>>n;
    vector<int> arr(n,0);
    for(int i=0;i<arr.size();i++){
        cin>>arr[i];
    }
    //vector<int> arr = {2,3,6,4,1};
    cout<<maximum02(arr,0)<<endl;
    cout<<minimum02(arr,0)<<endl;
    cout<<minimum01(arr,0);
   // display(arr);
}

void set1(){
   // cout<<fact(5);
   //cout<<power(2,10);
   //cout<<powerBtr(2,11);
}

void solve(){

    //cout<<"Hello from c++";
    //set1();
    set2();
}

int main(){

    solve();
    return 0;
}

// #include<bits/stdc++.h> 
// using namespace std; 
// #define MAX 100001 
  
// // dp[i] is going to store Log ( i !) in base 2 
// double dp[MAX]; 
  
// double probability(int n, int k) 
// { 
//     double ans = 0; // Initialize result 
  
//     // Iterate from k heads to n heads 
//     for (int i=k; i <= n; ++i) 
//     { 
//         double res = dp[n] - dp[i] - dp[n-i] - n; 
//         ans += pow(2.0, res); 
//     } 
  
//     return ans; 
// } 
  
// void precompute() 
// { 
//     // Preprocess all the logarithm value on base 2 
//     for (int i=2; i < MAX; ++i) 
//         dp[i] = log2(i) + dp[i-1]; 
// } 
  
// // Driver code 
// int main() 
// { 
//     precompute(); 
//     // int n,k;
//     // cin>>n;
//     // cin>>k;
//     // cout<<probability(n,k);
//     //Probability of getting 2 head out of 3 coins 
//     cout << probability(4, 3) << "\n"; 
  
//     //Probability of getting 3 head out of 6 coins 
//     cout << probability(6, 3) << "\n"; 
  
//     //Probability of getting 500 head out of 10000 coins 
//     cout << probability(1000, 500); 
  
//     return 0; 
// } 

File: Lecture-01-Recursion/Code.java
public class Code{
    public static void main(String[] args) {
        
        solve();
    }

    static void solve(){

        //set1();
        set2();
    }

    static void set1(){
        //System.out.println("Hello from java");
        System.out.println(Fact(5));

    }

    public static void set2(){
        // int[] arr=new int[10];  
        // for(int i=0;i<arr.length;i++){
        //   arr[i]=scn.nextInt();
        // }

        int[] arr={1,234,34,7,2,1,23,4,5,6,3,2,5,7,32};
        display(arr,0);
    }

    //-----------------------set1--------------

    public static int Fact(int n){
        if(n<=1) return 1;
 
        int smallAns=Fact(n-1);
        return smallAns*n;
 
         // return n<=1?1:Fact(n-1)*n;
     }
 
     public static int power(int a,int b){
       if(b==0) return 1;
 
       int smallAns=power(a,b-1);
       return smallAns*a;
     }
 
     public static int powerBtr(int a,int b){
         if(b==0) return 1;
 
         int smallAns=powerBtr(a,b/2);
         smallAns*=smallAns;
 
         return b%2==0?smallAns:smallAns*a;
     }
 
     public static int calls(int n) {  // n=5
         if (n <= 1) {
             System.out.println("base: " + n);
             return n + 1;
         }
 
         int count = 0;
 
         System.out.println("Pre: " + n);
         count += calls(n - 1);
 
         System.out.println("In: " + n);
 
         count += calls(n - 2);
         System.out.println("Post: " + n);
 
         return count + 3;
     }
 
     //set2.====================================================
     
     public static void display(int[] arr,int vidx){
         if(vidx==arr.length) return;
 
         System.out.print(arr[vidx] + " ");
         display(arr,vidx+1);
     }
 
     public static boolean find_01(int[] arr,int vidx,int data){
         if(vidx==arr.length) return false;
         
         if(arr[vidx]==data) return true;  // work in pre area.
         return find_01(arr,vidx+1,data);
     }
 
     
     public static boolean find_02(int[] arr,int vidx,int data){
         if(vidx==arr.length) return false;
         
         
         boolean ans=find_02(arr,vidx+1,data);
         if(ans) return true;   //work in post area.
         else
            return arr[vidx]==data;
     }
     
     
     public static int maximum(int[] arr,int vidx){
         if(vidx==arr.length-1) return arr[vidx];
        
         int maxele=maximum(arr,vidx+1);
         return Math.max(maxele,arr[vidx]); // max(a,b); for c++
     }
 
     static int max_=-1;
     public static void maximum_02(int[] arr,int vidx){
         if(vidx==arr.length-1) return ;
         
         max_=Math.max(arr[vidx],max_);
         maximum_02(arr,vidx);
     }
 
     public static int minimum(int[] arr,int vidx){
         if(vidx==arr.length-1) return arr[vidx];
        
         int maxele=minimum(arr,vidx+1);
         return Math.min(maxele,arr[vidx]);  // min(a,b); for c++
     }
}